### DISCUSSION TITLE
realtime: update when related field changed

### DISCUSSION THREAD

**Discussion by SunChero** [pocketbase/pocketbase#1502] [Q&A] [answered]
Hello,
as in the title, is there a way to receive realtime update event when a related collection changes ?
right now the related i can only receive update  when something  changes to the main collection or the relation pointers/id

similar to the expand query param we can pass to a getList() .. i couldnt find anything that can be passed to the subscribe() function to make it aware im also interested in changes to a related collection

or maybe my thinking is wrong ?

Thank you

**Comment by ganigeorgiev** [comment, marked as answer]
There is no support for triggering realtime events based on a relation record change (_the behavior would be kind of ambiguous and I'm not sure if we can implement it efficiently_).

You'll have to subscribe to the relation collection separately, either with the `*` topic or by using the id(s) of the related records. You can have as many `subscribe()` call as you want, internally the SDKs will try to maintain only a single SSE connection.

**Reply by dschissler** [reply]
> internally the SDKs will try to maintain only a single SSE connection.

How does that work?  Is there a slight delay before starting SSE connection so that the SDK can request all of the appropriate events all at once?

**Reply by ganigeorgiev** [reply]
@dschissler 

> How does that work?

We keep a single local [`EventSource`](https://developer.mozilla.org/en-US/docs/Web/API/EventSource) object per SDK client instance.

> Is there a slight delay before starting SSE connection so that the SDK can request all of the appropriate events all at once?

No. The SSE connection is just a long lived `GET /api/realtime` request.
The subscriptions are send via `POST /api/realtime` with body an array list of all topics that client has subscribed (eg. `["collection1/recordId1", "collection1/recordId2", "collection2/*"]`).
Keep in mind that the realtime subscriptions at the moment "reacts" only to events (create/update/delete) and there is no other requests fetching outside of the messages that are written to the response body of the long lived connection on record change.

**Reply by dschissler** [reply]
@ganigeorgiev It doesn't seem that you entirely answered my question.  If I subscribe three times in rapid succession then how is the SDK handling that?  The first subscription POST happens, and then what happens when the second POST is called?  Does it cancel the existing request?  Then what happens when the third request happens?  If I subscribe to ten things is that nine aborted requests?

**Reply by ganigeorgiev** [reply]
@dschissler 

> If I subscribe three times in rapid succession then how is the SDK handling that?

It depends when the subscribe calls happen.

If no SSE connection was established yet, then all subscriptions will be batched and send as one POST request.

If an SSE connection was already established, then each following subscribe sends individual POST request with the latest subscriptions state. The SDK handle the auto-cancellation for the realtime events gracefully and will wait only for the last submitted pending `POST /api/realtime`. The last submitted POST request contains all subscriptions (there are some async edge cases that we also handle, for example https://github.com/pocketbase/pocketbase/discussions/1150#discussioncomment-4285858).
You can find the JS SDK realtime implementation details in [RealtimeService.ts](https://github.com/pocketbase/js-sdk/blob/master/src/services/RealtimeService.ts).

**Reply by dschissler** [reply]
Thanks.  I'll definitely be going through soon.

**Reply by VittorioParagallo** [reply]
I just discovered that updates from related table (expand) are not streamed. So i found this discussion and IMO also updates from related table should be sent because, logically, the update is expected for the whole query, not just the main table. So i would call it bug more than a missing feature. Anyways is there any plan to develop this feature in the future?

**Reply by ganigeorgiev** [reply]
@VittorioParagallo No, I'm not planning working on this for now (see also https://github.com/pocketbase/pocketbase/discussions/4355).

> ... logically, the update is expected for the whole query, not just the main table.

No. The user subscribes either to a single record or a collection and if the record from that collection hasn't changed, aka. the relation id(s) remains the same, we simply don't trigger a realtime event. As mentioned above, listening to expanded records change could be ambiguous and I'm not sure if it can be implemented efficiently.
